### PR TITLE
SSCS-7812 Set DWP state on Direction Issued Submit for Resintatement agree/reject

### PR DESCRIPTION
**Pull request by danhamilton**
**Before creating a pull request make sure that:**

- [ ] commit messages are meaningful and follow good commit message guidelines
- [ ] README and other documentation has been updated / added (if needed)
- [ ] tests have been updated / new tests has been added (if needed)

Please remove this line and everything above and fill the following sections:

### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/SSCS-7812

### Change description ###

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ ] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/directionissued/DirectionIssuedAboutToSubmitHandler.java [hunk 1]
```diff
@@ -60,35 +64,128 @@ public boolean canHandle(CallbackType callbackType, Callback<SscsCaseData> callb
...
+        caseData.setInterlocReviewState(interlocReviewState);
+        caseData.setState(state);
+
+        return caseData;
+    }
+
+    @NotNull
+    private SscsCaseData updateCaseForDirectionType(CaseDetails<SscsCaseData> caseDetails, SscsCaseData caseData) {
+
+        if (DirectionType.PROVIDE_INFORMATION.toString().equals(caseData.getDirectionTypeDl().getValue().getCode())) {
```
Comment (ryandac) [inline src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/directionissued/DirectionIssuedAboutToSubmitHandler.java:103]: worth doing a null check for getDirectionTypeDl() and the following getValue

File: src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/directionissued/DirectionIssuedAboutToSubmitHandler.java [hunk 2]
```diff
@@ -60,35 +64,128 @@ public boolean canHandle(CallbackType callbackType, Callback<SscsCaseData> callb
...
+
+            caseData.setInterlocReviewState(AWAITING_INFORMATION.getId());
+
+        } else if (getPreValidStates().contains(caseDetails.getState())
+                && DirectionType.APPEAL_TO_PROCEED.toString().equals(caseData.getDirectionTypeDl().getValue().getCode())) {
```
Comment (ryandac) [inline src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/directionissued/DirectionIssuedAboutToSubmitHandler.java:108]: worth doing a null check for getDirectionTypeDl() and the following getValue

File: src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/directionissued/DirectionIssuedAboutToSubmitHandler.java [hunk 3]
```diff
@@ -60,35 +64,128 @@ public boolean canHandle(CallbackType callbackType, Callback<SscsCaseData> callb
...
+            caseData.setDateSentToDwp(LocalDate.now().toString());
+            caseData.setInterlocReviewState(AWAITING_ADMIN_ACTION.getId());
+
+        } else if (DirectionType.REFUSE_EXTENSION.toString().equals(caseData.getDirectionTypeDl().getValue().getCode())
+                && ExtensionNextEvent.SEND_TO_LISTING.toString().equals(caseData.getExtensionNextEventDl().getValue().getCode())) {
+            caseData = updateCaseAfterExtensionRefused(caseData, AWAITING_ADMIN_ACTION.getId(), State.RESPONSE_RECEIVED);
+
+        } else if (DirectionType.REFUSE_EXTENSION.toString().equals(caseData.getDirectionTypeDl().getValue().getCode())
```
Comment (ryandac) [inline src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/directionissued/DirectionIssuedAboutToSubmitHandler.java:116]: worth doing a null check for getDirectionTypeDl() and the following getValue

File: src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/directionissued/DirectionIssuedAboutToSubmitHandler.java [hunk 4]
```diff
@@ -60,35 +64,128 @@ public boolean canHandle(CallbackType callbackType, Callback<SscsCaseData> callb
...
+
+            caseData.setInterlocReviewState(AWAITING_INFORMATION.getId());
+
+        } else if (getPreValidStates().contains(caseDetails.getState())
+                && DirectionType.APPEAL_TO_PROCEED.toString().equals(caseData.getDirectionTypeDl().getValue().getCode())) {
+            caseData.setDateSentToDwp(LocalDate.now().toString());
+            caseData.setInterlocReviewState(AWAITING_ADMIN_ACTION.getId());
+
+        } else if (DirectionType.REFUSE_EXTENSION.toString().equals(caseData.getDirectionTypeDl().getValue().getCode())
+                && ExtensionNextEvent.SEND_TO_LISTING.toString().equals(caseData.getExtensionNextEventDl().getValue().getCode())) {
```
Comment (ryandac) [inline src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/directionissued/DirectionIssuedAboutToSubmitHandler.java:113]: worth doing a null check for getExtensionNextEventDl() and the following getValue

File: src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/directionissued/DirectionIssuedAboutToSubmitHandler.java [hunk 5]
```diff
@@ -60,35 +64,128 @@ public boolean canHandle(CallbackType callbackType, Callback<SscsCaseData> callb
...
+            caseData = updateCaseAfterExtensionRefused(caseData, AWAITING_ADMIN_ACTION.getId(), State.RESPONSE_RECEIVED);
+
+        } else if (DirectionType.REFUSE_EXTENSION.toString().equals(caseData.getDirectionTypeDl().getValue().getCode())
+                && ExtensionNextEvent.SEND_TO_VALID_APPEAL.toString().equals(caseData.getExtensionNextEventDl().getValue().getCode())) {
```
Comment (ryandac) [inline src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/directionissued/DirectionIssuedAboutToSubmitHandler.java:117]: worth doing a null check for getExtensionNextEventDl() and the following getValue

File: src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/directionissued/DirectionIssuedAboutToSubmitHandler.java [hunk 6]
```diff
@@ -60,35 +64,128 @@ public boolean canHandle(CallbackType callbackType, Callback<SscsCaseData> callb
...
+            caseData = updateCaseAfterExtensionRefused(caseData, null, State.WITH_DWP);
+
+        } else if (DirectionTypeItemList.GRANT_REINSTATEMENT.toString().equals(caseData.getDirectionTypeDl().getValue().getCode())
```
Comment (ryandac) [inline src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/directionissued/DirectionIssuedAboutToSubmitHandler.java:120]: worth doing a null check for getDirectionTypeDl() and the following getValue

File: src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/directionissued/DirectionIssuedAboutToSubmitHandler.java [hunk 7]
```diff
@@ -60,35 +64,128 @@ public boolean canHandle(CallbackType callbackType, Callback<SscsCaseData> callb
...
+    }
+
+    private SscsCaseData updateCaseAfterReinstatementGranted(SscsCaseData caseData) {
+
+        caseData.setReinstatementOutcome(ReinstatementOutcome.GRANTED);
+        caseData.setDwpState(DwpState.REINSTATEMENT_GRANTED.getId());
+
+        State previousState = caseData.getPreviousState();
+
+        if (! State.INTERLOCUTORY_REVIEW_STATE.getId().equals(previousState.getId())) {
```
Comment (ryandac) [inline src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/directionissued/DirectionIssuedAboutToSubmitHandler.java:141]: remove the space after the exclamation for readability

File: src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/directionissued/DirectionIssuedAboutToSubmitHandler.java [hunk 8]
```diff
@@ -129,16 +244,25 @@ public boolean canHandle(CallbackType callbackType, Callback<SscsCaseData> callb
...
         return sscsCaseDataPreSubmitCallbackResponse;
     }
 
-    private void updateCaseAfterExtensionRefused(SscsCaseData caseData, String interlocReviewState, State state) {
-        caseData.setHmctsDwpState("sentToDwp");
-        caseData.setDateSentToDwp(LocalDate.now().toString());
-        caseData.setInterlocReviewState(interlocReviewState);
-        caseData.setState(state);
+    private boolean shouldSetDwpState(SscsCaseData caseData) {
+        return ! reinstatementFeatureFlag
```
Comment (ryandac) [inline src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/directionissued/DirectionIssuedAboutToSubmitHandler.java:262]: remove space after the exclamation mark for readability

File: src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/directionissued/DirectionIssuedAboutToSubmitHandler.java [hunk 9]
```diff
@@ -129,16 +244,25 @@ public boolean canHandle(CallbackType callbackType, Callback<SscsCaseData> callb
...
+                || isNull(caseData.getReinstatementOutcome())
+                || (!caseData.getReinstatementOutcome().equals(ReinstatementOutcome.GRANTED)
+                && !caseData.getReinstatementOutcome().equals(ReinstatementOutcome.REFUSED));
```
Comment (ryandac) [inline src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/directionissued/DirectionIssuedAboutToSubmitHandler.java:265]: (!caseData.getReinstatementOutcome().equals(ReinstatementOutcome.GRANTED) 
&& !caseData.getReinstatementOutcome().equals(ReinstatementOutcome.REFUSED))

would be good to change the method name to the scenarios where (or perhaps where not in this case) the DwpState should be set.

File: src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/directionissued/DirectionIssuedAboutToSubmitHandler.java (hunk 1 above)
Comment (danhamilton) [reply src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/directionissued/DirectionIssuedAboutToSubmitHandler.java:103]: Already checked line 72

File: src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/directionissued/DirectionIssuedAboutToSubmitHandler.java [hunk 10]
```diff
@@ -60,35 +64,128 @@ public boolean canHandle(CallbackType callbackType, Callback<SscsCaseData> callb
...
+    private SscsCaseData updateCaseAfterExtensionRefused(SscsCaseData caseData, String interlocReviewState, State state) {
+        caseData.setHmctsDwpState("sentToDwp");
+        caseData.setDateSentToDwp(LocalDate.now().toString());
+        caseData.setInterlocReviewState(interlocReviewState);
+        caseData.setState(state);
+
+        return caseData;
+    }
+
+    @NotNull
```
[the remark this answers is not in the document]
Comment (danhamilton) [reply src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/directionissued/DirectionIssuedAboutToSubmitHandler.java:100]: I'll remove.

Is there a way to set this at a project level @ryandac? IntelliJ kindly adds it on extraction and I see it's value though not worth the cost of updating every method especially if we're agreed we should be keeping functions pure(ish) and always returning any mutated object

File: src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/directionissued/DirectionIssuedAboutToSubmitHandler.java (hunk 5 above)
Comment (ryandac) [reply src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/directionissued/DirectionIssuedAboutToSubmitHandler.java:115]: Is this null check done?

File: src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/directionissued/DirectionIssuedAboutToSubmitHandler.java [hunk 11]
```diff
@@ -107,19 +201,37 @@ public boolean canHandle(CallbackType callbackType, Callback<SscsCaseData> callb
...
+
 
         if (callback.getEvent() == EventType.DIRECTION_ISSUED) {
             footerService.createFooterAndAddDocToCase(url, caseData, DocumentType.DIRECTION_NOTICE,
-                Optional.ofNullable(caseData.getDateAdded()).orElse(LocalDate.now())
-                    .format(DateTimeFormatter.ofPattern("dd-MM-YYYY")),
-                caseData.getDateAdded(), null, documentTranslationStatus);
+                    Optional.ofNullable(caseData.getDateAdded()).orElse(LocalDate.now())
+                            .format(DateTimeFormatter.ofPattern("dd-MM-YYYY")),
+                    caseData.getDateAdded(), null, documentTranslationStatus);
```
Comment (ryandac) [inline src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/directionissued/DirectionIssuedAboutToSubmitHandler.java:221]: Formatting looks weird. Does adding the extra tab make it more readable?

File: src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/directionissued/DirectionIssuedAboutToSubmitHandler.java (hunk 5 above)
Comment (danhamilton) [reply src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/directionissued/DirectionIssuedAboutToSubmitHandler.java:115]: yes - see line 72 / previous comment

File: src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/directionissued/DirectionIssuedAboutToSubmitHandler.java (hunk 11 above)
Comment (danhamilton) [reply src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/directionissued/DirectionIssuedAboutToSubmitHandler.java:221]: could not be more readable :)

image

Comment (danhamilton) [reply src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/directionissued/DirectionIssuedAboutToSubmitHandler.java:221]: yes keeps it correctly aligned for reading

**Pull request closed without merging**